File: apps/tilawa/lib/features/islamic_widgets/domain/entities/wird_progress_widget_payload.dart
enum WirdWidgetTextDirection { ltr, rtl }

enum WirdWidgetAction { createPlan, openTodayWird, viewCompletedPlan }

final class WirdProgressWidgetPayload {
  WirdProgressWidgetPayload({
    required this.locale,
    required this.textDirection,
    required this.localizedTitle,
    required this.localizedSubtitle,
    required this.formattedAssignedAmount,
    required this.formattedCompletedAmount,
    required this.formattedRemainingAmount,
    required this.progressValue,
    required this.accessibilityLabel,
    required this.action,
    required this.generatedAt,
    required this.expiresAt,
    required this.isStale,
  }) : assert(
         progressValue >= 0 && progressValue <= 1,
         'progressValue must be between 0 and 1',
       ),
       assert(
         !expiresAt.isBefore(generatedAt),
         'expiresAt must not precede generatedAt',
       );

  static const int currentSchemaVersion = 1;

  int get schemaVersion => currentSchemaVersion;
  final String locale;
  final WirdWidgetTextDirection textDirection;
  final String localizedTitle;
  final String localizedSubtitle;
  final String formattedAssignedAmount;
  final String formattedCompletedAmount;
  final String formattedRemainingAmount;
  final double progressValue;
  final String accessibilityLabel;
  final WirdWidgetAction action;
  final DateTime generatedAt;
  final DateTime expiresAt;
  final bool isStale;

  Map<String, Object> toJson() => <String, Object>{
    'schemaVersion': schemaVersion,
    'locale': locale,
    'textDirection': textDirection.name,
    'localizedTitle': localizedTitle,
    'localizedSubtitle': localizedSubtitle,
    'formattedAssignedAmount': formattedAssignedAmount,
    'formattedCompletedAmount': formattedCompletedAmount,
    'formattedRemainingAmount': formattedRemainingAmount,
    'progressValue': progressValue,
    'accessibilityLabel': accessibilityLabel,
    'action': action.name,
    'generatedAt': generatedAt.toIso8601String(),
    'expiresAt': expiresAt.toIso8601String(),
    'isStale': isStale,
  };

  /// Tolerant decode at the widget-render boundary.
  ///
  /// Returns `null` when the payload cannot be trusted — an unknown
  /// [schemaVersion] major or any malformed field — so the caller shows the
  /// setup/no-data state instead of crashing (FR-041A1.9; Contract B invariant
  /// "Unknown schemaVersion major → render the setup/no-data state, never
  /// crash"). Well-formed current-version payloads round-trip [toJson].
  static WirdProgressWidgetPayload? tryParse(Map<String, Object?> json) {
    try {
      return WirdProgressWidgetPayload.fromJson(json);
    } on Object {
      return null;
    }
  }

  /// Strict decode of a current-schema payload. Throws [FormatException] on an
  /// unsupported [schemaVersion]; prefer [tryParse] at render boundaries.
  factory WirdProgressWidgetPayload.fromJson(Map<String, Object?> json) {
    final Object? version = json['schemaVersion'];
    if (version != currentSchemaVersion) {
      throw FormatException(
        'Unsupported wird widget schemaVersion: $version '
        '(expected $currentSchemaVersion)',
      );
    }
    return WirdProgressWidgetPayload(
      locale: json['locale']! as String,
      textDirection: WirdWidgetTextDirection.values.byName(
        json['textDirection']! as String,
      ),
      localizedTitle: json['localizedTitle']! as String,
      localizedSubtitle: json['localizedSubtitle']! as String,
      formattedAssignedAmount: json['formattedAssignedAmount']! as String,
      formattedCompletedAmount: json['formattedCompletedAmount']! as String,
      formattedRemainingAmount: json['formattedRemainingAmount']! as String,
      progressValue: (json['progressValue']! as num).toDouble(),
      accessibilityLabel: json['accessibilityLabel']! as String,
      action: WirdWidgetAction.values.byName(json['action']! as String),
      generatedAt: DateTime.parse(json['generatedAt']! as String),
      expiresAt: DateTime.parse(json['expiresAt']! as String),
      isStale: json['isStale']! as bool,
    );
  }
}

File: apps/tilawa/android/app/src/main/kotlin/com/tilawa/app/widget/wird/WirdProgressWidgetPayload.kt
package com.tilawa.app.widget.wird

import org.json.JSONObject

/** Text ordering the widget must lay out with. */
internal enum class WirdWidgetTextDirection(val wireName: String) {
    LTR("ltr"),
    RTL("rtl"),
    ;

    companion object {
        fun fromWireName(wireName: String): WirdWidgetTextDirection? =
            entries.firstOrNull { it.wireName == wireName }
    }
}

/**
 * Semantic tap action passed through from Spec 023. Concrete route resolution
 * (the `openKhatma` widget-bridge action) is deliberately deferred to
 * T-041A1-d; this type only decodes the intent.
 */
internal enum class WirdWidgetAction(val wireName: String) {
    CREATE_PLAN("createPlan"),
    OPEN_TODAY_WIRD("openTodayWird"),
    VIEW_COMPLETED_PLAN("viewCompletedPlan"),
    ;

    companion object {
        fun fromWireName(wireName: String): WirdWidgetAction? =
            entries.firstOrNull { it.wireName == wireName }
    }
}

/**
 * Parsed payload of a Daily Wird / Khatma progress snapshot (spec 041
 * amendment 041-A1, T-041A1-c decode half). The Kotlin counterpart of the
 * Flutter `WirdProgressWidgetPayload.tryParse`: the widget renders these
 * display-ready strings verbatim and performs no localization, digit
 * formatting, or plan math (Contract B invariant).
 *
 * Freshness (`generatedAt`/`expiresAt`/`isStale`) is intentionally omitted —
 * the enclosing [com.tilawa.app.widget.WidgetSnapshotEnvelope] owns
 * `generatedAtMs`/`validUntilMs`, and staleness is re-derived at render time.
 */
internal data class WirdProgressWidgetPayload(
    val locale: String,
    val textDirection: WirdWidgetTextDirection,
    val localizedTitle: String,
    val localizedSubtitle: String,
    val formattedAssignedAmount: String,
    val formattedCompletedAmount: String,
    val formattedRemainingAmount: String,
    val progressValue: Double,
    val accessibilityLabel: String,
    val action: WirdWidgetAction,
) {
    companion object {
        /**
         * Returns null when any required field is missing or invalid so the
         * caller renders the setup/no-data state instead of a broken frame
         * (FR-041A1.9). Mirrors the Flutter tolerant parser.
         */
        fun parse(payload: JSONObject): WirdProgressWidgetPayload? {
            val locale = payload.optString("locale")
            val textDirection = WirdWidgetTextDirection.fromWireName(
                payload.optString("textDirection"),
            ) ?: return null
            val title = payload.optString("localizedTitle")
            val subtitle = payload.optString("localizedSubtitle")
            val assigned = payload.optString("formattedAssignedAmount")
            val completed = payload.optString("formattedCompletedAmount")
            val remaining = payload.optString("formattedRemainingAmount")
            val accessibilityLabel = payload.optString("accessibilityLabel")
            val action = WirdWidgetAction.fromWireName(
                payload.optString("action"),
            ) ?: return null
            val progress = payload.optDouble("progressValue", Double.NaN)

            if (locale.isBlank() ||
                title.isBlank() ||
                subtitle.isBlank() ||
                assigned.isBlank() ||
                completed.isBlank() ||
                remaining.isBlank() ||
                accessibilityLabel.isBlank()
            ) {
                return null
            }
            if (progress.isNaN() || progress < 0.0 || progress > 1.0) return null

            return WirdProgressWidgetPayload(
                locale = locale,
                textDirection = textDirection,
                localizedTitle = title,
                localizedSubtitle = subtitle,
                formattedAssignedAmount = assigned,
                formattedCompletedAmount = completed,
                formattedRemainingAmount = remaining,
                progressValue = progress,
                accessibilityLabel = accessibilityLabel,
                action = action,
            )
        }
    }
}
